### PR TITLE
Replace P2P comms with RPC for Coinbase Monitoring Protocol

### PR DESCRIPTION
**Pull request by hansieodendaal**
## Description
This PR reworks the coinbase monitoring protocol to be in line with the latest wallet transaction validation protocol (PR#2614) and to make use of RPC comms rather than P2P comms. The two P2P request/response interactions to the base node, get chain metadata and fetch UTXOs, have been replaced by a single RPC call to query if the coinbase kernel, identified by the signature,
exists. This call also returns the best block height that is needed by the coinbase monitoring protocol to decide if the coinbase transaction should be cancelled or not. Specifically:
- Coinbase monitoring protocol making use of RPC comms (single call) rather than P2P comms (two request/response calls)
- Coinbase monitoring protocol taking into account new transactions statuses `Mined Unconfirmed` and `Mined Confirmed`
- Added two new unit tests
- Integration tests updated to reflect new transactions statuses

**Notes:** 
- Must be merged after #2614.
- Dependent on https://github.com/tari-project/wallet-grpc-client/pull/8 from `tari-project/wallet-grpc-client`

## Motivation and Context
This change brings the Coinbase Monitoring Protocol in line with #2614. In general using RPC is more efficient than the P2P comms, and also provided for some code simplifications.

## How Has This Been Tested?
- Tested with merge mining in Windows: base node + console wallet + merge mining proxy + XMRig.
- Tested with cucumber integration tests "Wallet Routing Mechanism" and "Transaction Info".
- Tested with two new unit tests.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [X] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:

* [X] I'm merging against the `development` branch.
* [X] I ran `cargo-fmt --all` before pushing.
* [X] I ran `cargo test` successfully before submitting my PR.
* [X] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [X] I have added tests to cover my changes.

### CODE REVIEW

**Comment by delta1** [comment]
> new transactions statuses Mined Confirmed and Mined Confirmed 

I think you mean Mined Unconfirmed for the first case 

> Two unit tests are failing, but they seem unrelated to this change. 

They are related to this change as they were passing in @philipr-za's commit. They probably need to be updated.

File: base_layer/core/src/base_node/rpc/service.rs
Comment (delta1) [inline base_layer/core/src/base_node/rpc/service.rs]: `chain_metadata`

File: base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs
Comment (delta1) [inline base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs:46]: why the change here?

**Comment by philipr-za** [comment]
> > new transactions statuses Mined Confirmed and Mined Confirmed
> 
> I think you mean Mined Unconfirmed for the first case
> 
> > Two unit tests are failing, but they seem unrelated to this change.
> 
> They are related to this change as they were passing in @philipr-za's commit. They probably need to be updated.

I believe these are fixed in the latest commit to https://github.com/tari-project/tari/pull/2614 and a lot of this PR currently consists of https://github.com/tari-project/tari/pull/2614 so I was going to wait until that got merged and this PR rebased to review fully.

File: base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs
Comment (delta1) [inline base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs:122]: The comment above probably needs to be updated to be kept up to date with these changes, at the very least regarding the fact that the protocol continues running until the coinbase is confirmed.

File: base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs
Comment (delta1) [inline base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs]: I know that the body is supposed to have at least 1 kernel, but this will panic instead of failing gracefully if that's not true. Do we already have mitigations in place that ensure the body will **always** have at least 1 kernel?

File: base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs
Comment (delta1) [inline base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs]: `metadata` is one word

File: base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs
Comment (delta1) [inline base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs]: You can either destructure the tuple here, or just return it: 

`Ok((utxo_found, metadata)) => (utxo_found, metadata),` or 

`Ok(response) => response,`

File: base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs
Comment (delta1) [inline base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs:456]: This function needs a comment just to describe what the return types indicate.

File: base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs
Comment (delta1) [inline base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs]: If this is important enough to not be `debug` then shouldn't it be `warn` instead of `info` ?

File: base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs
Comment (delta1) [inline base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs:532]: Minor nit: I don't think confirmed or unconfirmed need to be capitalised

**Review by delta1 — APPROVED** [review]
Looking good! Just a few comments

**Review by delta1 — CHANGES_REQUESTED** [review]
Sorry just requesting a few changes

**Comment by delta1** [comment]
> I believe these are fixed in the latest commit to #2614 and a lot of this PR currently consists of #2614 so I was going to wait until that got merged and this PR rebased to review fully.

Oh nice, sorry probably needs a rebase then and some of my comments may not apply

File: base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs
Comment (hansieodendaal) [reply base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs:46]: This is confusing in the log files, as the term "chain monitoring" is inclusive of both normal transactions and coinbase transactions.

Comment (delta1) [reply base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs:46]: Ah I see, sorry I misread it as just removing "protocol"

File: base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs
Comment (hansieodendaal) [reply base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs]: Good point - Implemented safe code.

File: base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs
Comment (hansieodendaal) [reply base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs]: Implemented the 2nd option, thanks.
  _(I do not think the first option improves the code)_

File: base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs
Comment (hansieodendaal) [reply base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs:532]: Kept the log message the same/similar as that for `transaction_broadcast_protocol`

**Comment by delta1** [comment]
Seems to be some compile errors

**Comment by hansieodendaal** [comment]
> Seems to be some compile errors

Thanks, they came about with the latest merge to dev tip, sorted.

File: base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs
Comment (philipr-za) [inline base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs]: This comment is for the old method so should be deleted with this update

File: base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs
Comment (philipr-za) [inline base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs:353]: Style Nit: Select! statements are hard to read and the automated formatting doesn't work in them. Generally I feel its best practice to keep as much code out of them as possible. Kind of feel that this code could easily have been inside the `query_coinbase_transaction(...)` function and kept this hard to read loop much cleaner and focused on the looping/retry aspect of this section of the code.

File: base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs
Comment (philipr-za) [inline base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs:456]: Referring to my Style Nit above. If that logic lived in this function then this over complicated return type is not needed.

File: base_layer/wallet/src/transaction_service/service.rs
Comment (philipr-za) [inline base_layer/wallet/src/transaction_service/service.rs]: This doesn't need to be async. There are no awaits in the code below.

**Review by philipr-za — DISMISSED** [review]
Other than the few Nits I left above this looks fine.

**Comment by delta1** [comment]
```
---- transaction_service::service::transaction_service_tx_broadcast stdout ----
thread 'transaction_service::service::transaction_service_tx_broadcast' panicked at 'assertion failed: `(left == right)`
  left: `Broadcast`,
 right: `Completed`', base_layer/wallet/tests/transaction_service/service.rs:4216:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by hansieodendaal** [comment]
> ```
> ---- transaction_service::service::transaction_service_tx_broadcast stdout ----
> thread 'transaction_service::service::transaction_service_tx_broadcast' panicked at 'assertion failed: `(left == right)`
>   left: `Broadcast`,
>  right: `Completed`', base_layer/wallet/tests/transaction_service/service.rs:4216:5
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ```

I believe this is an unrelated flaky test

File: base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs
Comment (hansieodendaal) [reply base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs:353]: Code moved as suggested

File: base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs
Comment (hansieodendaal) [reply base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs:456]: Looked at the logic; I feel it would overcomplicate `async fn query_coinbase_transaction` if the `if tip > self.block_height + self.resources.config.num_confirmations_required {` logic is moved in there as well, as that is what would be needed to change the return type to 
``` rust
Result<(bool, bool), TransactionServiceProtocolError>
```

**Comment by philipr-za** [comment]
> > ```
> > ---- transaction_service::service::transaction_service_tx_broadcast stdout ----
> > thread 'transaction_service::service::transaction_service_tx_broadcast' panicked at 'assertion failed: `(left == right)`
> >   left: `Broadcast`,
> >  right: `Completed`', base_layer/wallet/tests/transaction_service/service.rs:4216:5
> > note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> > ```
> 
> I believe this is an unrelated flaky test

I will look at this one in the morning

**Review by philipr-za — DISMISSED** [review]

**Pull request merged**
